plaidml/bridge/openvino/ops/prior_box.cpp

```diff
@@ -0,0 +1,410 @@
+  edsl::Tensor C_mix;
+  edsl::Tensor C_out;
+};
+
+PriorBoxImpl::PriorBoxImpl(const Context& context, const ngraph::op::PriorBoxAttrs& attributes)
+    : ctx(context), attrs(attributes) {
+  prepareConfig();
+  produceCenterBase();
+}
+
+// Preprocess data with some attributes
+void PriorBoxImpl::prepareConfig() {
+  auto* input_shape_ngraph_op = ngraph::as_type<ngraph::op::Constant>(ctx.layer->get_input_node_ptr(0));
+  std::vector<int> input_shape = input_shape_ngraph_op->get_vector<int>();
+  H = input_shape[0];
+  W = input_shape[1];
+  auto IHIW = ctx.operands.at(1);
+  IH = edsl::cast(op::slice(IHIW).add_dims({0}), DType::FLOAT32);
+  IW = edsl::cast(op::slice(IHIW).add_dims({1}), DType::FLOAT32);
+
+  num_of_prior = ngraph::op::PriorBox::number_of_priors(attrs);
+
+  aspect_ratios.push_back(1.0f);
+  for (const auto& aspect_ratio : attrs.aspect_ratio) {
+    bool exist = false;
+    for (const auto existed_value : aspect_ratios) exist |= std::fabs(aspect_ratio - existed_value) < 1e-6;
+
+    if (!exist) {
+      aspect_ratios.push_back(aspect_ratio);
+      if (attrs.flip) {
+        aspect_ratios.push_back(1.0f / aspect_ratio);
+      }
+    }
+  }
+  for (float ar : aspect_ratios) {
+    if (std::fabs(ar - 1.0f) < 1e-6) {
+      continue;
+    }
+    aspect_ratios_scale_size.push_back(std::sqrt(ar));
+  }
+
+  variance = attrs.variance;
+  if (variance.empty()) variance.push_back(0.1f);
+
+  step = edsl::cast(edsl::Tensor(attrs.step), DType::FLOAT32);
+  min_element_size = attrs.min_size.size();
+  TensorShape shape_ms(DType::FLOAT32, {min_element_size, 1});
+  Buffer buffer_ms(shape_ms);
+  buffer_ms.copy_from(attrs.min_size.data());
+  min_size = edsl::Constant(buffer_ms, "min_size");
+  if (!attrs.scale_all_sizes) {
+    // mxnet-like PriorBox
+    if (attrs.step == -1)
+      step = 1.f * IH / H;
+    else
+      step = step * IH;
+    min_size = min_size * IH;
+  }
+  edsl::Tensor a = IW / W;
+  edsl::Tensor b = step;
+  step_x = edsl::select(step == 0, IW / W, step);
+  step_y = edsl::select(step == 0, IH / H, step);
+
+  offset = attrs.offset;
+}
+
+void PriorBoxImpl::produceCenterBase() {
+  auto CW = edsl::cast(edsl::index({edsl::TensorDim(H), edsl::TensorDim(W), edsl::TensorDim(1)}, 1), DType::FLOAT32);
+  auto CH = edsl::cast(edsl::index({edsl::TensorDim(H), edsl::TensorDim(W), edsl::TensorDim(1)}, 0), DType::FLOAT32);
+  edsl::Tensor CW_normalized, CH_normalized;
+  CW_normalized = edsl::select(step == 0, (CW + 0.5f) * step_x / IW, (CW + offset) * step / IW);
+  CH_normalized = edsl::select(step == 0, (CH + 0.5f) * step_y / IH, (CH + offset) * step / IH);
+  C_mix = op::concatenate({CW_normalized, CH_normalized}, -1);
+}
+
+void PriorBoxImpl::processFixedSizePath() {
+  int fixed_size_count = attrs.fixed_size.size();
+  int fixed_ratio_count = attrs.fixed_ratio.size();
+  bool first_out = true;
+  size_t fixed_size_s = 0;
+  int density_s = 0, shift = 0;
+  float var = 0;
+
+  for (size_t s = 0; s < fixed_size_count; ++s) {
```

Comment:
If you've calculated lists of X-centers `CX` and Y-centers `CY`, each of length N, and a list of sizes `S` of length M, you can avoid this loop and the associated concatenation at line 262 and calculate all the box corners at once with something to the effect of
```
F1 = [1,0,1,0]
F2 = [0,1,0,1]
P = [-1,-1,1,1]
D = [W,H,W,H]
B[i,j,k] = (F1[i]*CX[j] + F2[i]*CY[j] + P[i]*S[k])/(2*D[i])
```

where the shape of B is 4xNxM (which can be reshaped into the correct flattened form).

Reply:
Also, an implementation detail for the last line of pseudocode:

Since you can't use explicit indices with elementwise operations, you will need to expand `F1`/`F2`/`P`/`D`/`S`/`CX`/`CY` to rank 3 tensors by adding size 1 dims, then rely on autobroadcasting. I.e., reshape `F1`, `F2`, and `P` to `(4, 1, 1)`, reshape `CX` and `CY` to `(1, N, 1)`, and reshape `S` to `(1, 1, M)`; then you can just write `B = (F1 * CX + F2 * CY + P * S) / (2 * D)` and autobroadcasting will handle the indices for you.

Reply:
Yes, i used to think about how to remove this loop, but i did not find a way to remove slice for the dynamic shape of each element so keep this path to avoid some duplicate calculations.  The main reason is for the attribute "density",  each element of fixed_size will use the density with its index, it makes each of them get a different shape for the reason of producing some elements of "density[i] * density[i]" size (different Center_temp for each attributes in one loop), and the concatenate for each attribute is a little different ( different reshape for each C_dst) .  One path like what you mentioned is to broadcast the center to the result size, but this may need sperate slice for each attribute or a concatenate of other attributes and may get some duplicate calculation for broadcast.   I may think more about this.

Reply:
Ah, thanks for that explanation, we had not understood the significance of `density`.

I notice that `density` is an optional attribute and I suspect that it is frequently not set. I think it would be worthwhile to have a forked path where the more efficient computation method is used where density is unset or uniform, and the current method is used if density varies.

Alternatively, perhaps it would be possible to first do all the computations for the first density (in the manner @jd-bartlett96  suggested) and then where there are multiple densities follow up & concatenate those results; then reorder everything to the correct order. I suspect this may not be possible to do efficiently (in particular I think that reordering step might be very messy), but if there is a reasonable way to do this I think it's a good idea.

Reply:
It's a good idea to isolate normal and special case.  I will try this, but it may need some time as each change shall pass the full tests.

Reply:
As I checked the reference implementation in openvino,  "density" will always have same count with "fixed_size" even openvion spec not shows this request, it seems to be inevitable to use density in all paths of "fixed_size", so I keep current logical.   I follow your advice to use auto broadcast to replace some hand writing ops,  but I do not find a way to use one formula to  combine all these conditions as the shape of each block is not equal, so there are still some concatenate and slice for the combination of results now.